### PR TITLE
Inlet Dirichlet boundary condition for tracer

### PR DESCRIPTION
**Pull request by oguevremont**
### Description

For some simulations, imposing Dirichlet conditions made the linear systems too stiff. We can now impose weakly boundary conditions at inlets using the velocity times Dirichlet value as a flux, in a Neumann approach.

### Testing

An application test is added.

### Documentation

The new type of boundary condition is added to the documentation.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [ ] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [ ] The PR description is cleaned and ready for merge

### CODE REVIEW

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
First batch of comments, I will continue tomorrow

File: applications_tests/lethe-fluid/tracer_inlet_dirichlet_bc.prm [hunk 1]
```diff
@@ -0,0 +1,121 @@
+set dimension = 2
```
Comment (blaisb) [inline applications_tests/lethe-fluid/tracer_inlet_dirichlet_bc.prm:1]: It would be better to use a test case for which you can have an analytical solution or something similar to verify that your implementation is good.

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 2]
```diff
@@ -102,8 +102,9 @@ For tracer boundary conditions, the defaults parameters are:
 
 * ``id`` is the number associated with the boundary condition. By default, Lethe assumes that the id is equivalent to the number of the bc.
 
-* ``type``: This is the type of boundary condition being imposed. At the moment, only dirichlet boundary conditions can be imposed for tracer.
-
+* ``type``: This is the type of boundary condition being imposed.
+    * ``dirichlet``, to impose a given tracer value at the boundary,
+    * ``inlet dirichlet``, to impose the Dirichlet boundary condition weakly using a Neumann approach.
```
[the remark this answers is not in the document]
Comment (blaisb) [reply doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:107]: or just inlet would be good. And then add a comment that explains what this is used for.
To be honest though, with the DG implementation, I don't think this boundary condition will be necessary.

File: include/solvers/tracer_assemblers.h [hunk 3]
```diff
@@ -33,6 +35,10 @@ template <int dim>
 class TracerAssemblerBase
 {
 public:
+  TracerAssemblerBase(std::shared_ptr<SimulationControl> p_simulation_control)
```
[the remark this answers is not in the document]
Comment (blaisb) [reply include/solvers/tracer_assemblers.h:38]: Yes agreed.
Also why does the AssemblerBase require the simulation control? Why not have the derived class require the simulation control instead?

File: include/solvers/tracer_assemblers.h [hunk 4]
```diff
@@ -57,13 +63,16 @@ class TracerAssemblerBase
   virtual void
   assemble_rhs(TracerScratchData<dim>    &scratch_data,
                StabilizedMethodsCopyData &copy_data) = 0;
+
+protected:
+  std::shared_ptr<SimulationControl> simulation_control;
```
Comment (blaisb) [inline include/solvers/tracer_assemblers.h:68]: I don't understand why you need to add that to the base class.

File: include/solvers/tracer_assemblers.h [hunk 5]
```diff
@@ -143,4 +154,51 @@ class TracerAssemblerBDF : public TracerAssemblerBase<dim>
...
+    const BoundaryConditions::TracerBoundaryConditions<dim>
+      &p_boundary_conditions_tracer)
+    : TracerAssemblerBase<dim>(simulation_control)
+    , boundary_conditions_tracer(p_boundary_conditions_tracer)
+  {}
+
+  /**
+   * @brief assemble_matrix Assembles the matrix
+   * @param scratch_data (see base class)
+   * @param copy_data (see base class)
```
Comment (blaisb) [inline include/solvers/tracer_assemblers.h:184]: add [in] [out]

File: include/solvers/tracer_scratch_data.h [hunk 6]
```diff
@@ -80,14 +80,25 @@ class TracerScratchData
                     const FiniteElement<dim>        &fe_tracer,
                     const Quadrature<dim>           &quadrature,
                     const Mapping<dim>              &mapping,
-                    const FiniteElement<dim>        &fe_fd)
+                    const FiniteElement<dim>        &fe_fd,
+                    const Quadrature<dim - 1>       &face_quadrature)
```
Comment (blaisb) [inline include/solvers/tracer_scratch_data.h:84]: I think face wuadrature could be after quadrature instead.

File: include/solvers/tracer_scratch_data.h [hunk 7]
```diff
@@ -206,6 +227,53 @@ class TracerScratchData
             this->laplacian_phi[q][k] = trace(this->hess_phi[q][k]);
           }
       }
+
+
+    // Arrays related to faces must be re-initialized for each cell, since they
+    // might depend on reference cell
+    // Only carry out this initialization if the cell is a boundary cell,
+    // otherwise these are wasted calculations
```
Comment (blaisb) [inline include/solvers/tracer_scratch_data.h:235]: This part will conflict hard with the DG branch...

File: include/solvers/tracer_scratch_data.h [hunk 8]
```diff
@@ -251,24 +319,66 @@ class TracerScratchData
...
+        Tensor<1, dim>                                  velocity_ale;
+        std::shared_ptr<Functions::ParsedFunction<dim>> velocity_ale_function =
+          ale.velocity;
+        Vector<double> velocity_ale_vector(dim);
+        for (unsigned int q = 0; q < n_q_points; ++q)
+          {
+            velocity_ale_function->vector_value(quadrature_points[q],
+                                                velocity_ale_vector);
+            for (unsigned int d = 0; d < dim; ++d)
+              velocity_ale[d] = velocity_ale_vector[d];
```
Comment (blaisb) [inline include/solvers/tracer_scratch_data.h:336]: I would make your impleentation compatible with the ALE also

File: source/solvers/tracer.cc [hunk 9]
```diff
@@ -31,6 +31,16 @@ Tracer<dim>::setup_assemblers()
 {
   this->assemblers.clear();
 
+  // Flux boundary condition
```
Comment (blaisb) [inline source/solvers/tracer.cc:34]: Instead of flux maybe inlet or neumann?

File: source/solvers/tracer_assemblers.cc [hunk 10]
```diff
@@ -319,3 +319,69 @@ TracerAssemblerBDF<dim>::assemble_rhs(TracerScratchData<dim>    &scratch_data,
...
+  TracerScratchData<dim>    &scratch_data,
+  StabilizedMethodsCopyData &copy_data)
+{
+  if (!scratch_data.is_boundary_cell)
+    return;
+
+  auto &local_rhs = copy_data.local_rhs;
+
+  // Flux boundary condition, loop on faces
+  // similar to deal.ii step-7
```
Comment (blaisb) [inline source/solvers/tracer_assemblers.cc:342]: I don't think this comment is necessary

**Review by blaisb — CHANGES_REQUESTED** [review]
Some comments.
Furthermore
1- I think you need a test with an analytical solution or something more firm to test that the implementation is correct
2- Are you sure this feature is really necessary? I have a feeling the DG branch would fix your issues better.
3- Have you tested that this feature actually fixes the issues you have right now with your problem?
